### PR TITLE
Adding validation for jsonExtractKey and jsonExtractScalar functions

### PR DESCRIPTION
**Pull request by fx19880617**
##  Description
Adding validation for json_extract_key and json_extract_scalar functions during sql compilation phase to avoid empty response

If user mistakenly put double quote on a the json key argument, say `jsonExtractScalar(myMapStr,"$.k1", 'STRING')`, then query parser recognizes the second argument `$.k1` as an identifier.  Then during segment pruning phase, the DataSchemaSegmentPruner prunes all the segments as there is no column matches `$.k1`. Then user will get an empty results.

This also means that we cannot error out this inside the corresponding functions, as these function won't be initialized at all.

Adding a function validation method during compilation time will allow us to find out the expression type mismatch then error out earlier and direct more comprehensive errors and guides to users.

Sample output would be:
image

### CODE REVIEW

**Review by Jackie-Jiang — COMMENTED** [review]
Why do we have this validation inside the parser? We should have a query validation phase after the query is parsed

**Comment by fx19880617** [comment]
> Why do we have this validation inside the parser? We should have a query validation phase after the query is parsed

The reason is that the current stack trace is very confusing and has no way for users to figure out what's wrong. Those identifiers and literals check will be much useful to help users correct their queries. 

Please let me know if you find some better place to put them.

**Comment by fx19880617** [comment]
> Why do we have this validation inside the parser? We should have a query validation phase after the query is parsed

Also it allows the query to fail fast at controller/broker not scatter-gather to servers.

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java [hunk 1]
```diff
@@ -902,11 +902,47 @@ private static Expression compileFunctionExpression(SqlBasicCall functionNode) {
...
       }
     }
+    validateFunction(functionName, operands);
     Expression functionExpression = RequestUtils.getFunctionExpression(functionName);
     functionExpression.getFunctionCall().setOperands(operands);
     return functionExpression;
   }
 
+  private static void validateFunction(String functionName, List<Expression> operands) {
+    switch (functionName.toUpperCase()) {
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java:912]: Remove underscore from the function name to match `json_extract_scalar` and `json_extract_key`. Please also add some test cases for function name with underscore

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java [hunk 2]
```diff
@@ -902,11 +902,47 @@ private static Expression compileFunctionExpression(SqlBasicCall functionNode) {
...
+    }
+  }
+
+  private static void validateJsonExtractScalarFunction(String functionName, List<Expression> operands) {
+    // Check that there are exactly 3 or 4 arguments
+    if (operands.size() < 3 || operands.size() > 4) {
+      throw new SqlCompilationException(
+          "Expected 3/4 arguments for transform function: jsonExtractScalar(jsonFieldName, 'jsonPath', 'resultsType', ['defaultValue'])");
+    }
+    if (!operands.get(1).isSetLiteral() || !operands.get(2).isSetLiteral()) {
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java:928]: Also validate the optional fourth argument?

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java [hunk 3]
```diff
@@ -902,11 +902,47 @@ private static Expression compileFunctionExpression(SqlBasicCall functionNode) {
...
+    }
+  }
+
+  private static void validateJsonExtractKeyFunction(String functionName, List<Expression> operands) {
+    // Check that there are exactly 3 or 4 arguments
+    if (operands.size() != 2) {
+      throw new SqlCompilationException(
+          "Exactly 2 arguments are required for transform function: jsonExtractKey(jsonFieldName, 'jsonPath')");
+    }
+    if (!operands.get(1).isSetLiteral() || !operands.get(2).isSetLiteral()) {
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java:940]: (Critical)
```suggestion
    if (!operands.get(1).isSetLiteral()) {
```

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java [hunk 4]
```diff
@@ -902,11 +902,47 @@ private static Expression compileFunctionExpression(SqlBasicCall functionNode) {
...
+          "Expected 3/4 arguments for transform function: jsonExtractScalar(jsonFieldName, 'jsonPath', 'resultsType', ['defaultValue'])");
+    }
+    if (!operands.get(1).isSetLiteral() || !operands.get(2).isSetLiteral()) {
+      throw new SqlCompilationException(
+          "Expected the second or third argument for transform function: jsonExtractScalar(jsonFieldName, 'jsonPath', 'resultsType', ['defaultValue']) to be a single-quoted literal value.");
+    }
+  }
+
+  private static void validateJsonExtractKeyFunction(String functionName, List<Expression> operands) {
+    // Check that there are exactly 3 or 4 arguments
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java:935]: Update the comment

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java [hunk 5]
```diff
@@ -902,11 +902,47 @@ private static Expression compileFunctionExpression(SqlBasicCall functionNode) {
...
+        break;
+      case "JSONEXTRACTKEY":
+        validateJsonExtractKeyFunction(functionName, operands);
+        break;
+    }
+  }
+
+  private static void validateJsonExtractScalarFunction(String functionName, List<Expression> operands) {
+    // Check that there are exactly 3 or 4 arguments
+    if (operands.size() < 3 || operands.size() > 4) {
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java:924]: (nit) For clarity
```suggestion
    if (operands.size() != 3 && operands.size() != 4) {
```

File: pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java [hunk 6]
```diff
@@ -902,11 +902,47 @@ private static Expression compileFunctionExpression(SqlBasicCall functionNode) {
...
+      throw new SqlCompilationException(
+          "Expected 3/4 arguments for transform function: jsonExtractScalar(jsonFieldName, 'jsonPath', 'resultsType', ['defaultValue'])");
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java:926]: `"Expect ..."`, same for other exception message

**Review by Jackie-Jiang — APPROVED** [review]
LGTM otherwise

**Pull request merged**